compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEUnrollAndDistributeWorkgroup.cpp

```diff
@@ -0,0 +1,511 @@
+          AMDAIE::LogicalObjectFifoFromMemrefOp target =
+              dmaOp.getTargetObjectFifo();
+          if (!operandMap.contains(source.getOutput())) {
+            rewriter.setInsertionPoint(source);
+            auto cloneOp = dyn_cast<AMDAIE::LogicalObjectFifoFromMemrefOp>(
+                rewriter.clone(*dmaOp.getSource().getDefiningOp()));
+            operandMap.map(source.getOutput(), cloneOp.getOutput());
+          }
+          if (!operandMap.contains(target.getOutput())) {
+            rewriter.setInsertionPoint(target);
+            auto cloneOp = dyn_cast<AMDAIE::LogicalObjectFifoFromMemrefOp>(
+                rewriter.clone(*dmaOp.getTarget().getDefiningOp()));
+            operandMap.map(target.getOutput(), cloneOp.getOutput());
+          }
+          builder.clone(*it, operandMap);
+        } else {
+          builder.clone(*it, operandMap);
+        }
+      }
+    }
+    return success();
+  }
+
+  LogicalResult matchAndRewrite(scf::ForOp forOp,
+                                PatternRewriter &rewriter) const override {
+    // Only unroll loops within a workgroup
+    if (!forOp->getParentOfType<AMDAIE::WorkgroupOp>()) return failure();
+
+    // Skip for ops with nested for ops. Wait until nested ones get resolved
+    // first.
+    auto nestedForOps = forOp.getOps<scf::ForOp>();
+    if (!nestedForOps.empty()) return failure();
+
+    // First, try to promote the for loop
+    if (succeeded(forOp.promoteIfSingleIteration(rewriter))) {
+      return success();
+    }
+
+    // Try hoisting dma ops outside the scf.for operation by sweeping once
+    // forward and once backward to hoist to before, respectively after the
+    // scf.for.
+    (void)hoistDmaOps<ForwardIterator>(rewriter, forOp);
+    (void)hoistDmaOps<ReverseIterator>(rewriter, forOp);
+
+    // Unroll and distribute the logical objectfifos
+    if (failed(loopUnrollAndDistributeLogicalObjectFifos(rewriter, forOp))) {
+      return failure();
+    }
+    return success();
+  }
+};
+
+/// Assign tiles to the logical objectfifos with local memory space (L1).
+/// The tiles are derived from the usage of the logical objectfifos within
+/// core operations, which are already assigned a tile location.
+LogicalResult assignLocalAieTiles(mlir::ModuleOp moduleOp) {
+  IRRewriter rewriter(moduleOp.getContext());
+
+  // Map from local objectfifos found to the tiles where they are used
+  DenseMap<AMDAIE::LogicalObjectFifoFromMemrefOp,
+           llvm::SmallSetVector<Value, 16>>
+      logicalObjectFifosToTiles;
+
+  // Utility function insert a local objectfifo - tile pair into the local
+  // objectfifo to tile map
+  auto insertTile = [&](AMDAIE::LogicalObjectFifoFromMemrefOp logicalObjectFifo,
+                        Value tileResult) -> void {
+    if (!logicalObjectFifosToTiles.contains(logicalObjectFifo)) {
+      llvm::SmallSetVector<Value, 16> tiles;
+      logicalObjectFifosToTiles[logicalObjectFifo] = tiles;
```

Comment:
Wouldn't `logicalObjectFifosToTiles[logicalObjectFifo] = {};` just work?

Reply:
Thanks, yes, this works.